src/main/scala/dev/kamu/cli/utility/DependencyGraph.scala

```diff
@@ -0,0 +1,25 @@
+package dev.kamu.cli.utility
+
+class DependencyGraph[A](dependsOn: A => List[A]) {
+  // TODO: Object creation should fail if the graph contains loops and report the loop as an error
+
+  def resolve(goal: List[A]): Seq[A] = {
+    type State = (Seq[A], Set[A]) // (dataset queue, queued datasets)
+    def add(x: A)(s: State): State = (s._1 :+ x, s._2 + x)
+    def isQueued(s: State)(x: A): Boolean = s._2.contains(x)
+    val emptyState = (Seq.empty[A], Set.empty[A])
+
+    @scala.annotation.tailrec
+    def queue(s: State, stack: List[A]): State = {
+      if (stack.nonEmpty) {
+        val deps = dependsOn(stack.head).reverse.filterNot(isQueued(s))
+        if (deps.isEmpty)
+          queue(add(stack.head)(s), stack.tail.dropWhile(isQueued(s)))
```

Comment:
this `dropWhile` confused me for a while, wouldn't it be more clear to check `if(!isQueued(stack.head))` as part of the outer `if` block and skip the item?

Reply:
Added if statement, removed dropWhile.